proposals/0024-astro-context.md

```diff
@@ -0,0 +1,139 @@
+
+In the client, in an Astro template file:
+
+```astro
+<html lang="en">
+	<head>
+		<meta charset="utf-8" />
+		<meta name="viewport" content="width=device-width" />
+		<title>Astro</title>
+	</head>
+	<body>
+		<p>Welcome, {Astro.context.user.name}</p>
+	</body>
+</html>
+```
+
+In the client, in an API endpoint:
+
+```typescript
+export function get({ params, context }) {
+  return context.user.name;
+}
+```
+
+In the server, using NodeJS SSR adapter:
+
+```typescript
+import { createServer } from 'http';
+
+const { handler } = await import('./path-to-entrypoint.mjs');
+
+server = createServer((req, res) => {
+  handler(req, res, { user: { name: 'Elvis' } });
+});
+
+server.listen();
+```
+
+## A bit more realistic example
+
+In the server, using NodeJS SSR adapter:
+
+```typescript
+import { createServer } from 'http';
+import express from 'express';
+import expressSession from 'express-session';
+import passport from 'passport';
+import { MyStrategy } from 'my-passport-strategy-library';
+
+const app = express();
+
+app.use(expressSession({ secret: 'parsley' }));
+
+passport.use(new MyStrategy());
+app.use(passport.initialize());
+app.use(passport.session());
+
+const { handler } = await import('./path-to-entrypoint.mjs');
+
+app.use((req, res, next) => {
+  handler(req, res, { user: req.user }, next);
+});
+
+server.listen();
+```
+
+# Motivation
+
+Using the NodeJS SSR adapter, it is already possible to integrate an Astro SSR
+website into a server using Express/Koa. The purpose of this change is to make
+it possible for state that is already derived by the server (such as
+authentication state) to easily be passed from Express/Koa to an Astro template
+via `Astro.context`, so that this information does not have to be re-derived
+from `Astro.request.headers`. 
+
+This allows users who already have existing infrastructure for deriving session
+state from a request to avoid re-implementing it to make it work in Astro
+templates, and allows the server to avoid paying the penalty of deriving that
+state twice per request.
+
+# Detailed design
+
+We would add a new parameter to the entrypoint exposed by the NodeJS entrypoint.
+This parameter is the context, which is available during rendering as
+`Astro.context` inside of `.astro` files, and as the `context` property on the
+argument given to API routes. This `context` can be anything, or
+`null`/`undefined` if it is not used. It is `null` when page is not being served
+using SSR, or when the page is being served through an SSR integration other
+than the NodeJS integration.
+
+This requires modifying the NodeJS adapter to have a compatible function
+signature, plumbing the `context` value through the SSR stack, and exposing it
+on the `Astro` faux-global. At the time of writing, a possible implementation is
+available in [astro#3934](https://github.com/withastro/astro/pull/3934).
+
+# Drawbacks
+
+This change, as described, would only be usable via the NodeJS adapter. Further
```

Comment:
Yeah, I think this is a bit of a problem. I think we would be hesitant to add a property to the Astro global that only worked in one adapter. If we had an idea to make it work for others that would make it more palpable.

In the past we have discussed ideas of middleware within Astro itself. Then presumably an Astro middleware could mutate a context object like you've described in this RFC.